hearpreprocess/tfds_pipeline.py

```diff
@@ -0,0 +1,247 @@
+        )
+        return builder
+
+    def run(self):
+        builder = self.get_tfds_builder()
+        # Download and prepare the data in the task folder
+        builder.download_and_prepare()
+        self.mark_complete()
+
+
+class ExtractTFDS(luigi_util.WorkTask):
+    """
+    Extracts the downloaded tfds dataset for a split
+    If a split is not present, data for that split will
+    be deterministically sampled from the train split, by the downstream
+    pipeline (Specifically `ExtractMetadata.split_train_test_val`)
+    """
+
+    outdir = luigi.Parameter()
+    split = luigi.Parameter()
+    download: DownloadTFDS = luigi.TaskParameter()
+
+    def requires(self):
+        return {"download": self.download}
+
+    @property
+    def output_path(self):
+        return self.workdir.joinpath(self.outdir)
+
+    @staticmethod
+    def load_tfds(builder, **as_dataset_kwargs) -> tf.data.Dataset:
+        """
+        This loads the dataset from the builder. Specifically this function returns
+        a dataset which will also contain the tfds_id to uniquely determine
+        each example in the dataset
+
+        https://github.com/tensorflow/datasets/blob/master/docs/determinism.ipynb
+        """
+        read_config = as_dataset_kwargs.pop("read_config", tfds.ReadConfig())
+        read_config.add_tfds_id = True  # Set `True` to return the 'tfds_id' key
+        return builder.as_dataset(read_config=read_config, **as_dataset_kwargs)
+
+    @staticmethod
+    def save_audio_labels(
+        dataset: tf.data.Dataset,
+        audio_dir: Path,
+        dataset_sample_rate: int,
+        audio_labels_df_path: Path,
+        label_idx_map: Dict[int, Any],
+    ):
+        """
+        Iterates over the tfds dataset and saves the audio as wavfile and
+        the audio labels as dataframe keyed on the tfds_id of the file.
+        The sample rate of the audio is required to save the numpy array
+        as wavfiles
+        """
+        filename_labels: Dict[str, Any] = {}
+        # Track label indices for all the audio files
+        all_label_idx: set = set()
+        for example in tqdm(tfds.as_numpy(dataset)):
+            # Only these - 'float32', 'float64', 'int16', 'int32' - formats
+            # are allowed in soundfile write function. If int64 type is found,
+            # convert to int32
+            numpy_audio = example["audio"]
+            if numpy_audio.dtype == np.int64:
+                numpy_audio = numpy_audio.astype("int32")
+            assert numpy_audio.dtype in [np.float32, np.float64, np.int16, np.int32], (
+                f"The audio's numpy array datatype: {numpy_audio.dtype} cannot be "
+                "saved with soundfile"
+            )
+
+            # Since the audio name is not available in tfds, the unique tfds_id
+            # is used to define the audio filename
+            # https://www.tensorflow.org/datasets/determinism
```

Comment:
Nice! I got stuck here when I was trying to make a pipeline with tfds a while back. Didn't find the tfds_id -- good stuff

Reply:
Yeah. It is unique, so maybe we use it.